### PR TITLE
Mygames

### PR DESCRIPTION
**Pull request by harthikss9**

### CODE REVIEW

File: lib/pages/history_page.dart [hunk 1]
```diff
@@ -1,25 +1,62 @@
...
+        builder: (context, snapshot) {
+          if (snapshot.connectionState == ConnectionState.waiting) {
+            return const Center(child: CircularProgressIndicator());
+          } else if (snapshot.hasError) {
+            return Center(child: Text("Error: ${snapshot.error}"));
+          } else if (!snapshot.hasData || snapshot.data!.isEmpty) {
+            return const Center(child: Text("No past events found."));
+          } else {
+            final events = snapshot.data!;
+            return ListView.builder(
```
Comment (jossean-rivera) [inline lib/pages/history_page.dart:40]: Looks pretty simplistic but fine.

File: lib/pages/my_games_page.dart [hunk 2]
```diff
@@ -1,25 +1,80 @@
...
+                        ScaffoldMessenger.of(context).showSnackBar(
+                          SnackBar(content: Text("Unregistered from ${event.name}")),
+                        );
+
+                        // Refresh the UI by calling setState
+                        setState(() {});
+                      },
+                      style: ElevatedButton.styleFrom(
+                        backgroundColor: Colors.red,
+                      ),
```
Comment (jossean-rivera) [inline lib/pages/my_games_page.dart:68]: Use theme. 

```suggestion
                      style: ElevatedButton.styleFrom(
                        backgroundColor: Theme.of(context).textTheme.error,
                      ),
```

You might need to add the import statement

File: lib/pages/sport_details_page.dart [hunk 3]
```diff
@@ -75,15 +76,22 @@ class _SportDetailsPageState extends State<SportDetailsPage> {
               itemCount: events.length,
               itemBuilder: (context, index) {
                 final event = events[index];
-                final hostUserRef = event.hostUserId as DocumentReference;
+
+                // Get current user ID
+                final currentUserId =
```
Comment (jossean-rivera) [inline lib/pages/sport_details_page.dart:81]: I'm sure this is gonna create conflicts with #19

File: lib/utilities/firebase_service.dart [hunk 4]
```diff
@@ -65,4 +76,110 @@ class FirebaseService {
...
+
+    // Fetch all events from the collection and filter manually
+    final eventsSnapshot = await _firestore.collection('events-collection').get();
+
+    final filteredEvents = eventsSnapshot.docs.where((doc) {
+      final data = doc.data() as Map<String, dynamic>;
+      final eventDateTime = data['dateTime'] as Timestamp;
+      final isFutureEvent = eventDateTime.toDate().isAfter(DateTime.now());
+      return eventIds.contains(doc.id) && (showFutureEvents ? isFutureEvent : !isFutureEvent);
+    }).toList();
```
Comment (jossean-rivera) [inline lib/utilities/firebase_service.dart:131]: Add the dateTime filter in the query to the database, do not load all into memory and then filter them in the app. Let the database do that. 
```dart
      final snapshot = await _firestore
          .collection('events-collection')
          .where('sportId', isEqualTo: sportId)
          .where('dateTime',
              isGreaterThan: Timestamp.now()) // Filter future events
          .get();
```

File: lib/utilities/firebase_service.dart [hunk 5]
```diff
@@ -65,4 +76,110 @@ class FirebaseService {
...
+
+    // Fetch all events and filter manually for past events
+    final eventsSnapshot = await _firestore.collection('events-collection').get();
+
+    final pastEvents = eventsSnapshot.docs.where((doc) {
+      final data = doc.data() as Map<String, dynamic>;
+      final eventDateTime = data['dateTime'] as Timestamp;
+      final isPastEvent = eventDateTime.toDate().isBefore(DateTime.now());
+      return eventIds.contains(doc.id) && isPastEvent;
+    }).toList();
```
Comment (jossean-rivera) [inline lib/utilities/firebase_service.dart:167]: Same thing. Include the date time in the database query.

File: lib/utilities/firebase_service.dart [hunk 6]
```diff
@@ -65,4 +76,110 @@ class FirebaseService {
...
+    print('Error fetching past events: $e');
+    return [];
+  }
+}
+
+
+
+
+
+
```
Comment (jossean-rivera) [inline lib/utilities/firebase_service.dart:184]:
```suggestion
```

**Review by jossean-rivera — CHANGES_REQUESTED** [review]

File: lib/utilities/firebase_service.dart (hunk 4 above)
Comment (vamsi633) [reply lib/utilities/firebase_service.dart:153]: used this code ,it ran fine and able to register , but getting error while fetching registered games in my games page , I looked in internet and came to know using "where" allows to filter for only 10 datas per each id

**Review by DeepakChowdary0811 — APPROVED** [review]

File: .gitignore [hunk 7]
```diff
@@ -43,4 +43,4 @@ app.*.map.json
 /android/app/release
 
 # Hide secrets
-secrets.dart
\ No newline at end of file
+/lib/secrets.dart
```
Comment (jossean-rivera) [inline .gitignore:46]:
```suggestion
secrets.dart
```

File: lib/pages/game_form.dart [hunk 8]
```diff
@@ -224,7 +226,7 @@ class _GameFormPageState extends State<GameFormPage> {
                         locationTitle: _locationTitle,
                         positionsRequired: [],
                         registeredUsers: [],
-                      ));
+                      ), userId);
```
Comment (jossean-rivera) [inline lib/pages/game_form.dart:229]: Why not set it on the properties of SportEvent? 
Like
```dart
SportEvent(
  sportId: widget.sportId,
  userId: userId)
```

File: lib/pages/my_games_page.dart [hunk 9]
```diff
@@ -1,18 +1,78 @@
...
   @override
   Widget build(BuildContext context) {
-    return Center(
-      child: FilledButton(
-        child: const Text("Go to History"),
-        onPressed: () {
-          context.goNamed(RouteNames.history);
+    final userId = widget.authenticationBloc.sportUser?.uuid ?? '';
+
+    return Scaffold(
```
Comment (jossean-rivera) [inline lib/pages/my_games_page.dart:25]: Do not use scaffold on the pages. The router is using the ScafoldWithNavBar as the shell of the rest of the app. That's why I removed them.

File: lib/pages/sport_details_page.dart [hunk 10]
```diff
@@ -132,6 +134,14 @@ class _SportDetailsPageState extends State<SportDetailsPage> {
                     itemCount: events.length,
                     itemBuilder: (context, index) {
                       final event = events[index];
+                      // Get current user ID
+                      final currentUserId =
+                          widget.authenticationBloc.sportUser?.uuid ?? '';
+
+                      // Check if the user is registered for this event
+                      final isRegistered =
+                          event.registeredUsers?.contains(currentUserId) ?? false;
```
Comment (jossean-rivera) [inline lib/pages/sport_details_page.dart:143]: This should have been done by checking the registration collection like we agreed on Tuesday last week. For a many-to-many relationship, you should use a join table like the registration collection. 

I guess for now it's fine, but this is a bad design.

File: lib/utilities/firebase_service.dart [hunk 11]
```diff
@@ -45,17 +45,28 @@ class FirebaseService {
...
-      print('User $userId registered for event $eventId');
-    } catch (e) {
-      print('Error registering for event: $e');
-    }
+  try {
+    final eventDoc = _firestore.collection('events-collection').doc(eventId);
+
+    // Add user to the event's registered users list and decrement available slots
+    await eventDoc.update({
+      'registeredUsers': FieldValue.arrayUnion([userId]),
```
Comment (jossean-rivera) [inline lib/utilities/firebase_service.dart:53]: If you have the registration collection, then this is unnecessary, you're saving the same data twice.

File: lib/utilities/firebase_service.dart [hunk 12]
```diff
@@ -64,13 +75,20 @@ class FirebaseService {
...
 
-      return null;
+      final hostlogDoc = _firestore.collection('host').doc();
+      await hostlogDoc.set({
+        'userId' : userId,
+        'eventId' : registrationDoc.id,
+    });
+
+          print('${userId} created the event ${registrationDoc.id}');
+
```
Comment (jossean-rivera) [inline lib/utilities/firebase_service.dart:91]: What is the host collection used for? You should simply use the registration collection, like we already agreed. If you want to get the events that the current user is hosting, then you can query the event collection where hostUserId equals the ID of the logged in user. 
Screenshot 2024-12-02 at 5 34 54 PM

**Review by jossean-rivera — CHANGES_REQUESTED** [review]
You're already querying the registration collection for getting the events a user is registered in getUserRegisteredEvents and you have the hostUserId in the events collection, so I don't see any need for another collection called hosts.

File: lib/pages/sport_details_page.dart (hunk 10 above)
Comment (vamsi633) [reply lib/pages/sport_details_page.dart:143]: it is checking the events collection table where registered users are stored

File: lib/pages/my_games_page.dart (hunk 9 above)
Comment (harthikss9) [reply lib/pages/my_games_page.dart]: done

File: .gitignore (hunk 7 above)
Comment (harthikss9) [reply .gitignore]: done

File: lib/pages/game_form.dart (hunk 8 above)
Comment (vamsi633) [reply lib/pages/game_form.dart]: resolved

**Review by jossean-rivera — APPROVED** [review]
It's good for now

**Pull request merged**